pkg/providers/cloudstack/cloudstack.go

```diff
@@ -492,71 +492,119 @@ func (p *cloudstackProvider) needsNewKubeadmConfigTemplate(workerNodeGroupConfig
-		log.V(4).Info("Old and new CloudStackMachineConfig Compute Offerings do not match", "machineConfig", generatedCsmc.Name, "oldComputeOffering", generatedCsmc.Spec.ComputeOffering,
-			"newComputeOffering", actualCsmc.Spec.ComputeOffering)
+
+	if !oldMachineConfig.Spec.DiskOffering.Equal(newMachineConfig.Spec.DiskOffering) {
+		log.V(4).Info(
+			"Old and new CloudStackMachineConfig DiskOffering does not match",
+			"machineConfig", oldMachineConfig.Name,
+			"oldDiskOffering", oldMachineConfig.Spec.DiskOffering,
+			"newDiskOffering", newMachineConfig.Spec.DiskOffering,
+		)
 		return true
 	}
-	if !generatedCsmc.Spec.DiskOffering.Equal(actualCsmc.Spec.DiskOffering) {
-		log.V(4).Info("Old and new CloudStackMachineConfig DiskOffering does not match", "machineConfig", generatedCsmc.Name, "oldDiskOffering", generatedCsmc.Spec.DiskOffering,
-			"newDiskOffering", actualCsmc.Spec.DiskOffering)
+
+	if !isEqualMap(oldMachineConfig.Spec.UserCustomDetails, newMachineConfig.Spec.UserCustomDetails) {
+		log.V(4).Info(
+			"Old and new CloudStackMachineConfig UserCustomDetails does not match",
+			"machineConfig", oldMachineConfig.Name,
+			"oldUserCustomDetails", oldMachineConfig.Spec.UserCustomDetails,
+			"newUserCustomDetails", newMachineConfig.Spec.UserCustomDetails,
+		)
 		return true
 	}
-	if len(generatedCsmc.Spec.UserCustomDetails) != len(actualCsmc.Spec.UserCustomDetails) {
-		log.V(4).Info("Old and new CloudStackMachineConfig UserCustomDetails does not match", "machineConfig", generatedCsmc.Name, "oldUserCustomDetails", generatedCsmc.Spec.UserCustomDetails,
-			"newUserCustomDetails", actualCsmc.Spec.UserCustomDetails)
+
+	if !isEqualMap(oldMachineConfig.Spec.Symlinks, newMachineConfig.Spec.Symlinks) {
+		log.V(4).Info(
+			"Old and new CloudStackMachineConfig Symlinks does not match",
+			"machineConfig", oldMachineConfig.Name,
+			"oldSymlinks", oldMachineConfig.Spec.Symlinks,
+			"newSymlinks", newMachineConfig.Spec.Symlinks,
+		)
 		return true
 	}
-	for key, value := range generatedCsmc.Spec.UserCustomDetails {
-		if value != actualCsmc.Spec.UserCustomDetails[key] {
-			log.V(4).Info("Old and new CloudStackMachineConfig UserCustomDetails does not match", "machineConfig", generatedCsmc.Name, "oldUserCustomDetails", generatedCsmc.Spec.UserCustomDetails,
-				"newUserCustomDetails", actualCsmc.Spec.UserCustomDetails)
-			return true
+
+	return false
+}
+
+func isEqualMap[K, V comparable](a, b map[K]V) bool {
+	if len(a) != len(b) {
+		return false
+	}
+
+	// Ensure all keys are present in b, and a's values equal b's values.
+	for k, av := range a {
+		if bv, ok := b[k]; !ok || av != bv {
+			return false
 		}
 	}
-	if len(generatedCsmc.Spec.Symlinks) != len(actualCsmc.Spec.Symlinks) {
-		log.V(4).Info("Old and new CloudStackMachineConfig Symlinks does not match", "machineConfig", generatedCsmc.Name, "oldSymlinks", generatedCsmc.Spec.Symlinks,
-			"newSymlinks", actualCsmc.Spec.Symlinks)
-		return true
+
+	return true
+}
```

Comment:
Yes. I decided against it because the comparison felt quite specific. I wanted to play with the contract for `V` to see if it can be made more flexible than simply `comparable`.